File: 2024/day6.py
def load_map(filename):
    with open(filename, 'r') as file:
        return [list(line.strip()) for line in file.readlines()]

def find_guard_start(grid):
    for y, row in enumerate(grid):
        for x, cell in enumerate(row):
            if cell in "^>v<":
                return (x, y), cell
    raise ValueError("Guard starting position not found")

def turn_right(direction):
    return {'^': '>', '>': 'v', 'v': '<', '<': '^'}[direction]

def move_forward(x, y, direction):
    moves = {'^': (0, -1), '>': (1, 0), 'v': (0, 1), '<': (-1, 0)}
    dx, dy = moves[direction]
    return x + dx, y + dy

def simulate_patrol(grid, start, direction, max_moves=1000):
    visited = set()
    x, y = start
    moves = 0
    state_history = []  # Tracks sequential states for cycle detection

    while moves < max_moves:
        state = (x, y, direction)
        if state in state_history:
            # Check for a cycle by ensuring the repeated state matches
            cycle_start = state_history.index(state)
            if state_history[cycle_start:] == state_history[:len(state_history[cycle_start:])]:
                print(f"Loop detected at ({x}, {y}) facing {direction}")
                return visited, True  # True loop detected
        state_history.append(state)

        visited.add((x, y))
        nx, ny = move_forward(x, y, direction)

        # Check boundaries or obstacles
        if not (0 <= ny < len(grid) and 0 <= nx < len(grid[0])) or grid[ny][nx] == '#':
            direction = turn_right(direction)  # Turn right if blocked
        else:
            x, y = nx, ny  # Move forward

        moves += 1

    print(f"Simulation completed. Visited positions: {len(visited)}")
    return visited, False

def find_valid_obstructions(grid, visited):
    valid_positions = set()
    for y, row in enumerate(grid):
        for x, cell in enumerate(row):
            if (x, y) in visited or cell != '.':
                continue

            # Add obstruction
            modified_grid = [list(row) for row in grid]
            modified_grid[y][x] = '#'

            # Re-simulate
            start, direction = find_guard_start(modified_grid)
            new_visited, has_loop = simulate_patrol(modified_grid, start, direction)

            # Check if the patrol is stuck in a loop
            if has_loop:
                valid_positions.add((x, y))
                print(f"Valid obstruction found at: ({x}, {y})")

    return valid_positions

def main():
    grid = load_map('day6-example.txt')
    print("Map loaded successfully:")
    for row in grid:
        print("".join(row))

    # Part One
    start, direction = find_guard_start(grid)
    visited, _ = simulate_patrol(grid, start, direction)
    print(f"Part One: {len(visited)}")

    # Part Two
    valid_obstructions = find_valid_obstructions(grid, visited)
    print(f"Part Two: {len(valid_obstructions)}")

if __name__ == "__main__":
    main()

File: 2024/day6.js
const fs = require("fs");

const MAX_MOVES = 1000;

// Load the map from the input file
function loadMap(filename) {
    const data = fs.readFileSync(filename, "utf-8").trim();
    return data.split("\n").map(row => row.split(""));
}

// Find the guard's starting position and direction
function findGuardStart(map) {
    for (let y = 0; y < map.length; y++) {
        for (let x = 0; x < map[0].length; x++) {
            if ("^>v<".includes(map[y][x])) {
                return { x, y, direction: map[y][x] };
            }
        }
    }
    throw new Error("Guard starting position not found!");
}

// Simulate guard movement
function simulateGuard(map) {
    const directions = { "^": [0, -1], ">": [1, 0], "v": [0, 1], "<": [-1, 0] };
    const turnRight = { "^": ">", ">": "v", "v": "<", "<": "^" };

    const visited = new Set();
    let { x, y, direction } = findGuardStart(map);
    let moves = 0;

    while (true) {
        const key = `${x},${y}`;
        if (!visited.has(key)) visited.add(key);

        const [dx, dy] = directions[direction];
        const nx = x + dx;
        const ny = y + dy;

        if (
            nx < 0 || ny < 0 || nx >= map[0].length || ny >= map.length ||
            map[ny][nx] === "#"
        ) {
            // Turn right if hitting a boundary or obstacle
            direction = turnRight[direction];
        } else {
            // Move forward
            x = nx;
            y = ny;
        }

        moves++;
        if (x < 0 || y < 0 || x >= map[0].length || y >= map.length) break; // Guard exits map
        if (moves > MAX_MOVES) break; // Prevent infinite loop
    }

    return visited;
}

// Test obstructions and identify valid ones for loops
function testObstructions(map) {
    const originalVisited = simulateGuard(map);
    const obstructions = [];

    for (let y = 0; y < map.length; y++) {
        for (let x = 0; x < map[0].length; x++) {
            if (map[y][x] !== ".") continue; // Skip non-empty cells
            if (originalVisited.has(`${x},${y}`)) continue; // Skip original patrol path

            const testMap = map.map(row => [...row]);
            testMap[y][x] = "#"; // Add an obstruction

            const testVisited = simulateGuard(testMap);
            if (testVisited.size < originalVisited.size) {
                obstructions.push({ x, y });
            }
        }
    }

    return obstructions;
}

// Main function
function main(filename) {
    const map = loadMap(filename);
    console.log("Map loaded successfully:");
    console.log(map.map(row => row.join("")).join("\n"));

    // Part One
    const visited = simulateGuard(map);
    console.log("Part One:", visited.size);

    // Part Two
    const obstructions = testObstructions(map);
    console.log("Part Two:", obstructions.length);
}

// Run the program
main("day6-example.txt");
